File: Android/_setup/_ProjectName_/src/main/java/com/twist/dataengine/entities/TM_Region.java
package com.twist.dataengine.entities;

import com.google.gson.Gson;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by Twist Mobile on 8/25/2016.
 */

public class TM_Region {

    public boolean regionsLoaded = false;

    //public boolean isNode = false;
    public String type;
    //public String parentId;
    public String id;
    public String title;
    public String parentId; //new headache
    private List<TM_Region> children = new ArrayList<>();

    private TM_Region(String type, String id, String title, TM_Region parentRegion) {
        this.type = type;
        //this.parentId = parentId;
        this.id = id;
        this.title = title;
        if (parentRegion != null) {
            this.parentId = parentRegion.id;
            //this.isNode = false;
            parentRegion.children.add(this);
        } else {
            //this.isNode = true;
            this.parentId = "";
        }
        regionList.add(this);
    }

    private static List<TM_Region> regionList = new ArrayList<>();

    public static TM_Region getRegion(String type, String id, String title, TM_Region parent) {
        for (TM_Region region : regionList) {
            if (region.type.equals(type) && region.id.equals(id) && (region.parentId == null || parent == null || region.parentId.equals(parent.id))) { // new headache
                return region;
            }
        }
        return new TM_Region(type, id, title, parent);
    }

    public static TM_Region getRegionFromAll(String type, String title) {
        for (TM_Region region : regionList) {
            if (region.type.equals(type) && region.title.equals(title)) {
                return region;
            }
        }
        return new TM_Region(type, title, title, null);
    }

    public static List<TM_Region> getRegions(TM_Region parent) {
        if (parent != null) {
            return parent.children;
        } else {
            List<TM_Region> regions = new ArrayList<>();
            for (TM_Region region : regionList) {
                if (region.parentId == null || region.parentId.length() == 0) {
                    regions.add(region);
                }
            }
            return regions;
        }
    }

    @Override
    public String toString() {
        return this.title;
    }

    @Override
    public boolean equals(Object object) {
        if (object instanceof TM_Region) {
            TM_Region other = (TM_Region) object;
            return (this.type.equals(other.type) && this.id.equals(other.id));
        }
        return super.equals(object);
    }

    public String toJson() {
        return new Gson().toJson(this);
    }

    public static TM_Region fromJson(String jsonString) {
        return new Gson().fromJson(jsonString, TM_Region.class);
    }
}

File: IOS/TMStore/TMStore/Classes/Services/TMShippingEngine/TMRegion.m
//
//  TMRegion.m
//  TMShippingSDK
//
//  Created by Rishabh Jain on 28/09/16.
//  Copyright © 2016 Twist Mobile. All rights reserved.
//

#import "TMRegion.h"
static NSMutableArray* regionList = nil;
@implementation TMRegion

- (id)initWithoutAppendingInRegionList {
    self = [super init];
    if (self) {
        _regionChildren = [[NSMutableArray alloc] init];
        _regionId = @"";
        _regionIsLoaded = false;
        _regionIsNode = false;
        _regionTitle = @"";
        _regionType = @"";
    }
    return self;
}
- (id)init {
    self = [super init];
    if (self) {
        if (regionList == nil) {
            regionList = [[NSMutableArray alloc] init];
        }
        [regionList addObject:self];
        
        
        _regionChildren = [[NSMutableArray alloc] init];
        _regionId = @"";
        _regionIsLoaded = false;
        _regionIsNode = false;
        _regionTitle = @"";
        _regionType = @"";
    }
    return self;
}
- (id)init:(NSString*)regionType regionId:(NSString*)regionId regionTitle:(NSString*)regionTitle regionParent:(TMRegion*)regionParent {
    self = [super init];
    if (self) {
        if (regionList == nil) {
            regionList = [[NSMutableArray alloc] init];
        }
        [regionList addObject:self];
        
        
        _regionChildren = [[NSMutableArray alloc] init];
        _regionId = regionId;
        _regionTitle = regionTitle;
        _regionType = regionType;
        _parentRegion = regionParent;
        if (regionParent != nil) {
            _regionIsNode = false;
            [regionParent.regionChildren addObject:self];
        } else {
            _regionIsNode = true;
        }
        _regionIsLoaded = false;
    }
    return self;
}
+ (TMRegion*)getRegion:(NSString*)regionType regionId:(NSString*)regionId regionTitle:(NSString*)regionTitle regionParent:(TMRegion*)regionParent {
    for (TMRegion* region in regionList) {
        if ([regionType isEqualToString:region.regionType] && [regionId isEqualToString:region.regionId] && regionParent == region.parentRegion) {
            return region;
        }
    }
    return [[TMRegion alloc] init:regionType regionId:regionId regionTitle:regionTitle regionParent:regionParent];
}
+ (TMRegion*)getRegionFromAll:(NSString*)regionType regionTitle:(NSString*)regionTitle {
    for (TMRegion* region in regionList) {
        if ([regionType isEqualToString:region.regionType] && [regionTitle isEqualToString:region.regionTitle]) {
            return region;
        }
    }
    return [[TMRegion alloc] init:regionType regionId:regionTitle regionTitle:regionTitle regionParent:nil];
}
+ (NSMutableArray*)getRegions:(TMRegion*)regionParent {
    if (regionParent != nil) {
        return regionParent.regionChildren;
    } else {
        NSMutableArray* regions = [[NSMutableArray alloc] init];
        for (TMRegion* region in regionList) {
            if (region.regionIsNode) {
                [regions addObject:region];
            }
        }
        return regions;
    }
}
+ (TMRegion*)findRegionFromId:(NSString*)regionId regionType:(NSString*)regionType regionParent:(TMRegion*)regionParent {
    for (TMRegion* region in regionList) {
        if (regionParent == nil) {
            if ([region.regionId isEqualToString:regionId] && [region.regionType isEqualToString:regionType]) {
                return region;
            }
        } else {
            if ([region.regionId isEqualToString:regionId] && [region.regionType isEqualToString:regionType] && [region.parentRegion.regionId isEqualToString:regionParent.regionId]) {
                return region;
            }
        }
    }
    return nil;
}

//@Override
//public String toString() {
//    return this.title;
//}
//
//@Override
//public boolean equals(Object object) {
//    if (object instanceof TMRegion) {
//        TMRegion other = (TMRegion) object;
//        return (this.type.equals(other.type) && this.id.equals(other.id));
//    }
//    return super.equals(object);
//}
//
//public String toJson() {
//    return new Gson().toJson(this);
//}
//
//public static TMRegion fromJson(String jsonString) {
//    return new Gson().fromJson(jsonString, TMRegion.class);
//}
@end
